bin/python/combine_reports_from_date_range.py

```diff
@@ -0,0 +1,109 @@
+            tests_table.add_row([test, *[str(result.get(label, 0)) for label in result_labels]])
+    else:
+        tests_table.add_row(['No results'])
+
+    category_label_sets = map(lambda category: set(category.keys()), results['categories'].values())
+    result_labels = _reduce_labels(category_label_sets)
+    categories_table = MdTable(['categories'] + result_labels)
+    if results['categories']:
+        for category, result in results['categories'].items():
+            row_result = [str(result.get(label, 0)) for label in result_labels]
+            categories_table.add_row([category, *row_result])
+    else:
+        categories_table.add_row(['No results'])
+
+    missing_table = MdTable(['missing tests', 'count'])
+    if results['missing']:
+        for test in sorted(results['missing'].keys()):
+            missing_table.add_row([test, str(results['missing'][test])])
+    else:
+        missing_table.add_row(['None'])
+
+    return "\n%s\n" * 3 % (tests_table.render(), categories_table.render(), missing_table.render())
+
+def _get_local_reports(device, reports_dir, start, end):
+    LOGGER.info('Looking for reports locally')
+    prog = re.compile(r'^report_%s_\d{4}-\d{2}-\d{2}T\d{6}\+\d{4}.*\.json$' % device)
+    ts = re.compile(r'\d{4}-\d{2}-\d{2}T\d{6}\+\d{4}')
+    json_files = [f for f in os.listdir(reports_dir) if prog.match(f)]
+    for json_file in json_files:
+        timestamp = ts.search(json_file).group(0) 
+        start_str = start.isoformat().replace(':', '') if start else None
+        end_str = end.isoformat().replace(':', '') if end else None
+        if (start_str and timestamp < start_str) or (end_str and timestamp > end_str):
+            LOGGER.info('Skipping file %s' % json_file)
+            continue
+        LOGGER.info('Processing file %s' % json_file)
+        with open(os.path.join(reports_dir, json_file), 'r') as json_file_handler:
+            yield json.loads(json_file_handler.read())
+
+def main(device, start=None, end=None, gcp=None, reports_dir=os.path.join('inst', 'reports')):
+    aggregate = {'tests': {}, 'categories': {}, 'missing': {}}
+    device = device.replace(':', '').lower()
+    if gcp:
+        device_full = ":".join([device[i:i + 2] for i in range(0, len(device), 2)])
+        json_reports = gcp.get_reports_from_date_range(device_full, start=start, end=end)
+    else:
+        json_reports = _get_local_reports(device, reports_dir, start, end)
+
+    for json_report in json_reports:
+        for test in json_report.get('missing_tests', []):
+            aggregate['missing'].setdefault(test, 0)
+            aggregate['missing'][test] += 1
+        for _, module_result in json_report.get('modules', {}).items():
+            for test, test_result in module_result.get('tests', {}).items():
+                result = test_result.get('result')
+                category = test_result.get('category')
+                aggregate['tests'].setdefault(test, {})
```

Comment:
NB: these can be chained, although I'm not convinced it's better (more readable), e.g.
aggregate['tests'].setdefault(test, {}).setdefault(result,0)

Up to you if you want to change or not.